### PR TITLE
feat(security): add ReplicaSession interceptor 

### PR DESCRIPTION
**Pull request by levy5307**
* add ReplicaSession hook to deal with some situations, such as session connected、sending a message.
* add some configurations

### CODE REVIEW

File: src/main/java/com/xiaomi/infra/pegasus/rpc/async/SecurityReplicaSessionHook.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    try {
+      loginContext.login();
+    } catch (LoginException le) {
+      logger.error("authentication failed", le);
+      System.exit(-1);
+    }
+
+    subject = loginContext.getSubject();
+    logger.info("login succeed, as user {}", subject.getPrincipals().toString());
+  }
```
[the remark this answers is not in the document]
Comment (levy5307) [reply src/main/java/com/xiaomi/infra/pegasus/rpc/async/SecurityReplicaSessionHook.java:39]: The javax.security.auth.login.LoginContext class provides the basic methods used to authenticate subjects, and provides a way to develop an application independent of the underlying authentication technology. The LoginContext consults a Configuration to determine the authentication services, or LoginModule(s), configured for a particular application. Therefore, different LoginModules can be plugged in under an application without requiring any modifications to the application itself.
(Ref)[https://docs.oracle.com/javase/7/docs/technotes/guides/security/jaas/JAASRefGuide.html#LoginContext]

File: configuration/pegasus_jaas.conf [hunk 2]
```diff
@@ -0,0 +1,10 @@
+client {
+	com.sun.security.auth.module.Krb5LoginModule required
```
Comment (neverchanje) [inline configuration/pegasus_jaas.conf:2]: Does this config format support comment? Can you write a comment of description on top of each config item?
Furthermore, does it mean that every of our users should have this file in their project?

File: src/main/java/com/xiaomi/infra/pegasus/rpc/async/ReplicaSessionHookManager.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package com.xiaomi.infra.pegasus.rpc.async;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ReplicaSessionHookManager {
```
Comment (neverchanje) [inline src/main/java/com/xiaomi/infra/pegasus/rpc/async/ReplicaSessionHookManager.java:6]: I would suggest naming this class a `SessionInterceptor` to keep aligned with `TableIntercetor`. And you can place this class in `interceptor/` package.

**Review by Shuo-Jia — APPROVED** [review]

File: src/main/java/com/xiaomi/infra/pegasus/rpc/async/ReplicaSessionHookManager.java (hunk 3 above)
Comment (levy5307) [reply src/main/java/com/xiaomi/infra/pegasus/rpc/async/ReplicaSessionHookManager.java:6]: Yes, I can put it in `interceptor/`, but there would be two interceptorManager, which are `InterceptorManager` and `ReplicaSessionInterceptorManager`

File: configuration/pegasus_jaas.conf (hunk 2 above)
Comment (levy5307) [reply configuration/pegasus_jaas.conf:2]: 1.I have tested that this config does't support comment. So I can write a file later to show what`s the meaning of these config item.
2. Yes, our users should have this file in their project.

Comment (levy5307) [reply configuration/pegasus_jaas.conf:3]: Yes, every our users should have this file in their project

File: src/main/java/com/xiaomi/infra/pegasus/rpc/async/ClusterManager.java [hunk 4]
```diff
@@ -56,7 +58,12 @@ public ClusterManager(ClientOptions opts) throws IllegalArgumentException {
     replicaGroup = getEventLoopGroupInstance(opts.getAsyncWorkers());
     metaGroup = getEventLoopGroupInstance(1);
     tableGroup = getEventLoopGroupInstance(1);
-    enableAuth = opts.isEnableAuth();
+    if (opts.isEnableAuth()) {
+      ReplicaSessionInterceptor hook =
+          new SecurityReplicaSessionInterceptor(
+              opts.getJaasConf(), opts.getServiceName(), opts.getServiceFQDN());
+      ReplicaSessionInterceptorManager.instance().addHook(hook);
```
Comment (Shuo-Jia) [inline src/main/java/com/xiaomi/infra/pegasus/rpc/async/ClusterManager.java:65]: Just as your previous commet about `TableInterceptor`: https://github.com/XiaoMi/pegasus-java-client/pull/125#discussion_r481045604, the `ReplicaSessionInterceptor` can be put into `ReplicaSessionInterceptorManager`

File: configuration/pegasus_jaas.conf (hunk 2 above)
Comment (levy5307) [reply configuration/pegasus_jaas.conf]: I removed the `pegasus_jaas.conf`. So we can add it in later pull request. And I will refactor it in a better way at that time.

File: src/main/java/com/xiaomi/infra/pegasus/rpc/async/ReplicaSession.java [hunk 5]
```diff
@@ -39,10 +40,13 @@
   }
 
   public ReplicaSession(
-      rpc_address address, EventLoopGroup rpcGroup, int socketTimeout, boolean enableAuth) {
```
[the remark this answers is not in the document]
Comment (levy5307) [reply src/main/java/com/xiaomi/infra/pegasus/rpc/async/ReplicaSession.java:42]: I don't think we should create a `ReplicaSessionInterceptorManager` for each `ReplicationSession`. Each ClusterManager has a `ReplicaSessionInterceptorManager` is enough

**Review by neverchanje — APPROVED** [review]

**Review by Shuo-Jia — APPROVED** [review]

**Pull request merged**